python/python_direct/python_translate.cpp

```diff
@@ -0,0 +1,506 @@
+  }
+
+ private:
+  PythonTranslator(std::ostream& os, Fusion* fusion)
+      : printer_(os), fusion_(fusion) {}
+
+  bool isScheduledTensorView(TensorView* tv) const {
+    NVF_ERROR(tv != nullptr);
+    const std::vector<IterDomain*>& logical = tv->domain()->logical();
+    const std::vector<IterDomain*>& loop = tv->domain()->loop();
+    // short-circuit: check same length
+    if (logical.size() != loop.size()) {
+      return true;
+    }
+
+    for (size_t idx : c10::irange(logical.size())) {
+      if (logical.at(idx) != loop.at(idx)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  // Check that all of the expression's inputs are defined in FusionDefinition.
+  bool checkExpressionDependencies(Expr* e) {
+    // TODO Add check_view_dependency
+    return std::all_of(
+        e->inputs().begin(), e->inputs().end(), [&](const Val* v) {
+          return visited_vals_.count(v) > 0;
+        });
+  }
+
+  void translate() {
+    printer_.generateFusionDefinition();
+
+    // Add Fusion inputs to FusionDefinition
+    for (nvfuser::Val* v : fusion_->inputs()) {
+      dispatch(v);
+    }
+
+    // Gather all expressions in CPP Fusion.
+    const std::vector<nvfuser::Expr*> fusion_exprs = fusion_->exprs();
+    std::deque<nvfuser::Expr*> to_visit(
+        fusion_exprs.begin(), fusion_exprs.end());
+
+    // TODO: Scalar expressions are not handled by Fusion::exprs, so gather them
+    // manually.
+
+    // Topological search of Fusion expressions
+    size_t skip_count = 0;
+    std::unordered_set<nvfuser::Expr*> visited;
+    while (!to_visit.empty()) {
+      Expr* e = to_visit.front();
+      to_visit.pop_front();
+
+      NVF_ERROR(
+          skip_count <= to_visit.size(),
+          "Cycle detected: None of the expressions can be processed!");
+
+      // short-circuit: skip if already visited
+      if (visited.count(e) > 0) {
+        continue;
+      }
+
+      // TODO: short-circuit: skip Split and Merge expressions created by
+      // Reshape
+      // TODO: short-circuit: skip Resize expressions created by Slice
+
+      bool is_expr_inputs_valid =
+          std::all_of(e->inputs().begin(), e->inputs().end(), [this](Val* v) {
+            return !v->isA<TensorView>() ||
+                !isScheduledTensorView(v->as<TensorView>());
+          });
+      NVF_ERROR(
+          is_expr_inputs_valid,
+          "Found a TensorView with scheduled loop domain.");
+
+      // Handle scalars and constants not generated by separate expression.
+      std::vector<Val*> scalars;
+      std::copy_if(
+          e->inputs().begin(),
+          e->inputs().end(),
+          std::back_inserter(scalars),
+          [](Val* v) { return v->isScalar(); });
+      std::for_each(scalars.begin(), scalars.end(), [this](const Val* v) {
+        dispatch(v);
+      });
+
+      // short-circuit: add to back of stack if not all of the expression's
+      // dependencies are satisfied.
+      if (!checkExpressionDependencies(e)) {
```

Comment:
should we assert here instead? Since Fusion::exprs() returns toplogically sorted expressions. you shouldn't need to maintain the traverse order.